packages/devtools_app/lib/src/charts/treemap.dart

```diff
@@ -0,0 +1,517 @@
+import 'package:flutter/material.dart';
+
+import '../trees.dart';
+import '../utils.dart';
+
+class Treemap extends StatefulWidget {
+  const Treemap({
+    @required this.rootNode,
+    @required this.levelsVisible,
+    @required this.width,
+    @required this.height,
+    @required this.onTap,
+  });
+
+  final TreemapNode rootNode;
+
+  /// The depth of children visible from this Treemap widget.
+  ///
+  /// For example, levelsVisible = 2 at root Treemap:
+  /// ```
+  /// _______________
+  /// |     Root    |
+  /// ---------------
+  /// |    l = 1    |
+  /// |  ---------  |
+  /// |  | l = 2 |  |
+  /// |  |       |  |
+  /// |  |       |  |
+  /// |  ---------  |
+  /// ---------------
+  /// ```
+  final int levelsVisible;
+
+  final double width;
+
+  final double height;
+
+  final VoidCallback onTap;
+
+  @override
+  _TreemapState createState() => _TreemapState();
+}
+
+enum PivotType { pivotByMiddle, pivotBySize }
+
+class _TreemapState extends State<Treemap> {
+  static const double minHeightToDisplayText = 20.0;
+  PivotType pivotType = PivotType.pivotBySize;
+
+  TreemapNode rootNode;
+
+  @override
+  void initState() {
+    super.initState();
+    rootNode = widget.rootNode;
+  }
+
+  bool get shouldDisplayText => widget.height > minHeightToDisplayText;
+
+  void cellOnTap(TreemapNode child) {
+    if (child != null) {
+      setState(() {
+        rootNode = child;
```

Comment:
can you add a todo to push rootNode state management into the controller responsible for managing the data? Ideally we'd like the memory controller to maintain the root node so that we can update the root node for the heap table accordingly. We'll want to do something like this for the code size page as well.

Reply:
Done